graylog2-server/src/main/java/org/graylog2/database/PaginatedDbService.java

```diff
@@ -0,0 +1,220 @@
+        return save.getSavedObject();
+    }
+
+    /**
+     * Deletes the {@link DTO} for the given ID from the database.
+     *
+     * @param id ID of the {@link DTO} to delete
+     */
+    public void delete(String id) {
+        db.removeById(new ObjectId(id));
+    }
+
+    /**
+     * Returns a {@link PaginatedList<DTO>} for the given query and pagination parameters.
+     * <p>
+     * This method is only accessible by subclasses to avoid exposure of the {@link DBQuery} and {@link DBSort}
+     * interfaces to consumers.
+     *
+     * @param query the query to execute
+     * @param sort the sort builder for the query
+     * @param page the page number that should be returned
+     * @param perPage the number of entries per page
+     * @return the paginated list
+     */
+    protected PaginatedList<DTO> findPaginatedWithQueryAndSort(DBQuery.Query query, DBSort.SortBuilder sort, int page, int perPage) {
+        final DBCursor<DTO> cursor = db.find(query)
+                .sort(sort)
+                .limit(perPage)
+                .skip(perPage * Math.max(0, page - 1));
+
+        return new PaginatedList<>(asImmutableList(cursor), cursor.count(), page, perPage);
+    }
+
+    private ImmutableList<DTO> asImmutableList(Iterator<? extends DTO> cursor) {
+        return ImmutableList.copyOf(cursor);
+    }
+
+    /**
+     * Returns a {@link PaginatedList<DTO>} for the given query, filter and pagination parameters.
+     * <p>
+     * Since the database cannot execute the filter function directly, this method streams over the result cursor
+     * and executes the filter function for each database object. This is more expensive than using
+     * {@link PaginatedDbService#findPaginatedWithQueryFilterAndSort(DBQuery.Query, Predicate, DBSort.SortBuilder, int, int) #findPaginatedWithQueryAndSort()}.
+     * <p>
+     * This method is only accessible by subclasses to avoid exposure of the {@link DBQuery} and {@link DBSort}
+     * interfaces to consumers.
+     *
+     * @param query the query to execute
+     * @param filter the filter to apply to each database entry
+     * @param sort the sort builder for the query
+     * @param page the page number that should be returned
+     * @param perPage the number of entries per page
+     * @return the paginated list
+     */
+    protected PaginatedList<DTO> findPaginatedWithQueryFilterAndSort(DBQuery.Query query,
+                                                                     Predicate<DTO> filter,
+                                                                     DBSort.SortBuilder sort,
+                                                                     int page,
+                                                                     int perPage) {
+        final AtomicInteger total = new AtomicInteger(0);
+
+        // First created a filtered stream
+        final Stream<DTO> dtoStream = streamQueryWithSort(query, sort)
```

Comment:
What's the impact of the intermediary stream (and the following flat-mapping) on memory consumption?
I could imagine that peeking into each result item (and this loading every document from MongoDB) has some effect on the overall memory consumption.

Could it make sense to execute two MongoDB queries here, one returning the total number of documents matching the query and one returning the requested page with limit and offset in the MongoDB query itself instead of being filtered afterwards?

Reply:
And do we require the total number of documents in every case or could this be an additional database request if required?

Reply:
Yes, this will increase memory consumption because it needs to iterate over all objects. The main reason we do this is not the counting of total objects but to execute the filter `Predicate` for each object returned by the `DBQuery`.

One example where we have to use a filter `Predicate` are permission checks, where we query a list of objects from the database and need to filter that list for objects the user has access to.
In that case we cannot use the `limit()` and `skip()` methods on the `DBCursor` but have to do that in memory to get correct total counts and page counts.

Consumers should use `findPaginatedWithQueryAndSort()` if possible to avoid this overhead. But in some cases there is no way around that.